File: src/memory/builder/flow_unit.py
"""Navigator flow-unit helpers for Ariad Builder delivery."""

from __future__ import annotations

from dataclasses import dataclass

from memory.builder.delivery_cursor import (
    BuilderDeliveryCursor,
    get_delivery_cursor,
    set_delivery_cursor,
)
from memory.builder.lifecycle_ribbon import render_lifecycle_ribbon
from memory.builder.surface_protocol import wrap_ariad_surface
from memory.storage.store import Store

FLOW_UNIT_STORY_BY_STORY = "story_by_story"
FLOW_UNIT_DELIVERY_STORY = "delivery_story"
ALLOWED_FLOW_UNITS = (FLOW_UNIT_STORY_BY_STORY, FLOW_UNIT_DELIVERY_STORY)


@dataclass(frozen=True)
class NavigatorFlowUnitReport:
    journey: str
    method: str
    active_item: str | None
    active_item_title: str | None
    active_item_level: str | None
    flow_unit: str
    source: str
    cursor: BuilderDeliveryCursor


def effective_navigator_flow_unit(cursor: BuilderDeliveryCursor) -> tuple[str, str]:
    """Return the effective Navigator flow unit and its source."""
    if cursor.navigator_flow_unit in ALLOWED_FLOW_UNITS:
        return cursor.navigator_flow_unit, "cursor"
    return FLOW_UNIT_STORY_BY_STORY, "default"


def set_navigator_flow_unit(
    store: Store,
    *,
    journey: str,
    method: str,
    flow_unit: str,
) -> NavigatorFlowUnitReport:
    """Persist the Navigator-facing flow unit for active Builder delivery."""
    if flow_unit not in ALLOWED_FLOW_UNITS:
        allowed = ", ".join(ALLOWED_FLOW_UNITS)
        raise ValueError(f"navigator flow unit must be one of: {allowed}")
    existing = get_delivery_cursor(store, journey)
    if existing is None:
        raise ValueError("delivery cursor is required before choosing navigator flow unit")
    updated = set_delivery_cursor(
        store,
        journey=journey,
        method=method,
        active_item=existing.active_item,
        active_item_title=existing.active_item_title,
        active_item_level=existing.active_item_level,
        active_checkpoint=existing.active_checkpoint,
        pending_confirmation=existing.pending_confirmation,
        last_delivery_event="navigator_flow_unit_selected",
        cadence_profile=existing.cadence_profile,
        cadence_limits=existing.cadence_limits,
        granularity_decision=existing.granularity_decision,
        navigator_flow_unit=flow_unit,
        child_work_items=existing.child_work_items,
        aggregate_checkpoint_status=existing.aggregate_checkpoint_status,
    )
    return NavigatorFlowUnitReport(
        journey=journey,
        method=method,
        active_item=updated.active_item,
        active_item_title=updated.active_item_title,
        active_item_level=updated.active_item_level,
        flow_unit=flow_unit,
        source="cursor",
        cursor=updated,
    )


def inspect_navigator_flow_unit(
    store: Store,
    *,
    journey: str,
    method: str,
) -> NavigatorFlowUnitReport:
    """Inspect the effective Navigator flow unit for active Builder delivery."""
    cursor = get_delivery_cursor(store, journey)
    if cursor is None:
        raise ValueError("delivery cursor is required before inspecting navigator flow unit")
    flow_unit, source = effective_navigator_flow_unit(cursor)
    return NavigatorFlowUnitReport(
        journey=journey,
        method=method,
        active_item=cursor.active_item,
        active_item_title=cursor.active_item_title,
        active_item_level=cursor.active_item_level,
        flow_unit=flow_unit,
        source=source,
        cursor=cursor,
    )


def render_navigator_flow_unit_report(report: NavigatorFlowUnitReport) -> str:
    """Render a deterministic Ariad surface for the Navigator flow-unit decision."""
    body = "\n".join(
        [
            "Delivery",
            render_lifecycle_ribbon("expand"),
            "",
            "╭────────────────────────────────────────────────────────╮",
            "│        🧭■  NAVIGATOR FLOW UNIT                        │",
            "│                                                        │",
            _card_text("journey"),
            _card_text(report.journey),
            "│                                                        │",
            _card_text("method"),
            _card_text(report.method),
            "│                                                        │",
            _card_text("active item"),
            _card_text(report.active_item or "none"),
            "│                                                        │",
            _card_text("active item title"),
            *_card_wrapped(report.active_item_title or "none"),
            "│                                                        │",
            _card_text("active item level"),
            _card_text(report.active_item_level or "none"),
            "│                                                        │",
            _card_text("effective flow unit"),
            _card_text(report.flow_unit),
            "│                                                        │",
            _card_text("source"),
            _card_text(report.source),
            "│                                                        │",
            _card_text("available choices"),
            *_card_prefixed(
                (
                    "story_by_story: child User/Technical Stories remain Navigator-facing lifecycle units",
                    "delivery_story: parent Delivery Story becomes the Navigator-facing lifecycle unit while child stories remain traceable Driver work packages",
                ),
                "○",
            ),
            "│                                                        │",
            _card_text("default"),
            *_card_wrapped(
                "story_by_story preserves current Ariad Builder behavior when no choice is recorded."
            ),
            "│                                                        │",
            _card_text("boundary"),
            *_card_wrapped(
                "No Plan, implementation, validation, push, or release work was executed."
            ),
            "╰────────────────────────────────────────────────────────╯",
        ]
    )
    return wrap_ariad_surface("navigator_flow_unit", body + "\n")


def _card_text(text: str) -> str:
    width = 54
    return f"│ {text[:width]:<{width}} │"


def _card_prefixed(items: tuple[str, ...], prefix: str) -> list[str]:
    lines: list[str] = []
    for item in items:
        wrapped = _wrap_plain_text(item, width=52)
        for index, line in enumerate(wrapped):
            marker = prefix if index == 0 else " "
            lines.append(_card_text(f"{marker} {line}"))
    return lines or [_card_text("none")]


def _card_wrapped(text: str) -> list[str]:
    return [_card_text(line) for line in _wrap_plain_text(text, width=54)]


def _wrap_plain_text(text: str, *, width: int) -> list[str]:
    words = text.split()
    lines: list[str] = []
    current = ""
    for word in words:
        if len(word) > width:
            if current:
                lines.append(current)
                current = ""
            for start in range(0, len(word), width):
                lines.append(word[start : start + width])
            continue
        candidate = f"{current} {word}".strip()
        if len(candidate) > width and current:
            lines.append(current)
            current = word
        else:
            current = candidate
    if current:
        lines.append(current)
    return lines or ["none"]

File: ts/src/builder/flowUnit.ts
// CV22.DS7.US8 — the Navigator flow unit.
//
// Port of `src/memory/builder/flow_unit.py`. Plateau 3 took the read side, because
// Plan needs the EFFECTIVE flow unit before it can record story authority (bounded
// story authority exists only under `story_by_story`). Plateau 5 completes it: the
// write side was re-sequenced out of Scope F by the panel, since the DS smoke opens
// with `set-flow-unit` and seeding the flow unit by a raw cursor write instead
// would be the unrecorded mutation plateau 3a ruled out.
//
// The selection surface is CHOSEN by the unit, not fixed: `delivery_story` renders
// DELIVERY_STORY_SCOPE_CONFIRMATION over the child work packages, `story_by_story`
// renders NEXT_STORY_CONFIRMATION over the first child as the recommended story.
// One function, two surface ids — a port that hard-codes either one fails the
// corpus guard that compares each recorded id against its own wrapped marker.

import type { WritableDatabase } from "#db/database.ts";
import { pyStrip } from "#util/pythonText.ts";
import { cardPrefixed, cardText, cardWrapped } from "./card.ts";
import {
  type BuilderDeliveryCursor,
  type CursorWriteDeps,
  getDeliveryCursor,
  setDeliveryCursor,
} from "./deliveryCursor.ts";
import { wrapAriadSurface } from "./surfaceProtocol.ts";

export const FLOW_UNIT_STORY_BY_STORY = "story_by_story";
export const FLOW_UNIT_DELIVERY_STORY = "delivery_story";
export const ALLOWED_FLOW_UNITS = [FLOW_UNIT_STORY_BY_STORY, FLOW_UNIT_DELIVERY_STORY] as const;

/**
 * Python `effective_navigator_flow_unit`.
 *
 * An unset or UNRECOGNIZED value falls back to `story_by_story` with source
 * `default`, rather than raising — so a cursor written by a newer engine, or one
 * carrying a typo, still yields a usable flow unit instead of blocking the
 * lifecycle. The source is returned because the Navigator-facing surfaces
 * distinguish "you chose this" from "this is the default".
 */
export function effectiveNavigatorFlowUnit(cursor: BuilderDeliveryCursor): {
  flowUnit: string;
  source: string;
} {
  const current = cursor.navigatorFlowUnit;
  if (current !== null && (ALLOWED_FLOW_UNITS as readonly string[]).includes(current)) {
    return { flowUnit: current, source: "cursor" };
  }
  return { flowUnit: FLOW_UNIT_STORY_BY_STORY, source: "default" };
}

export interface NavigatorFlowUnitReport {
  readonly journey: string;
  readonly method: string;
  readonly activeItem: string | null;
  readonly activeItemTitle: string | null;
  readonly activeItemLevel: string | null;
  readonly flowUnit: string;
  readonly source: string;
  readonly cursor: BuilderDeliveryCursor;
}

/**
 * Python `set_navigator_flow_unit`.
 *
 * Validation BEFORE the cursor read, so an unknown unit reports the allowed values
 * even when no cursor exists. Everything else on the cursor carries forward
 * untouched: choosing a flow unit is not a lifecycle transition, and the only
 * fields it writes are the unit itself and `last_delivery_event`.
 */
export function setNavigatorFlowUnit(
  db: WritableDatabase,
  options: { journey: string; method: string; flowUnit: string },
  deps: CursorWriteDeps,
): NavigatorFlowUnitReport {
  if (!(ALLOWED_FLOW_UNITS as readonly string[]).includes(options.flowUnit)) {
    throw new Error(`navigator flow unit must be one of: ${ALLOWED_FLOW_UNITS.join(", ")}`);
  }
  const existing = getDeliveryCursor(db, options.journey);
  if (existing === null) {
    throw new Error("delivery cursor is required before choosing navigator flow unit");
  }
  const updated = setDeliveryCursor(
    db,
    {
      journey: options.journey,
      method: options.method,
      activeItem: existing.activeItem,
      activeItemTitle: existing.activeItemTitle,
      activeItemLevel: existing.activeItemLevel,
      activeCheckpoint: existing.activeCheckpoint,
      pendingConfirmation: existing.pendingConfirmation,
      lastDeliveryEvent: "navigator_flow_unit_selected",
      cadenceProfile: existing.cadenceProfile,
      cadenceLimits: existing.cadenceLimits,
      granularityDecision: existing.granularityDecision,
      navigatorFlowUnit: options.flowUnit,
      childWorkItems: existing.childWorkItems,
      aggregateCheckpointStatus: existing.aggregateCheckpointStatus,
    },
    deps,
  );
  return {
    journey: options.journey,
    method: options.method,
    activeItem: updated.activeItem,
    activeItemTitle: updated.activeItemTitle,
    activeItemLevel: updated.activeItemLevel,
    flowUnit: options.flowUnit,
    source: "cursor",
    cursor: updated,
  };
}

/** Python `inspect_navigator_flow_unit`. */
export function inspectNavigatorFlowUnit(
  db: WritableDatabase,
  options: { journey: string; method: string },
): NavigatorFlowUnitReport {
  const cursor = getDeliveryCursor(db, options.journey);
  if (cursor === null) {
    throw new Error("delivery cursor is required before inspecting navigator flow unit");
  }
  const { flowUnit, source } = effectiveNavigatorFlowUnit(cursor);
  return {
    journey: options.journey,
    method: options.method,
    activeItem: cursor.activeItem,
    activeItemTitle: cursor.activeItemTitle,
    activeItemLevel: cursor.activeItemLevel,
    flowUnit,
    source,
    cursor,
  };
}

/** Python `_active_delivery`. */
function activeDelivery(report: NavigatorFlowUnitReport): string {
  if (report.activeItem === null) return "none";
  const title = report.activeItemTitle ? ` — ${report.activeItemTitle}` : "";
  return `🟦[${report.activeItem}]${title}`;
}

/** Python `_scope_item_lines`: an EMPTY list renders nothing, not `none`. */
function scopeItemLines(label: string, items: readonly string[]): string[] {
  if (items.length === 0) return [];
  return [
    "│                                                        │",
    cardText(label),
    ...cardPrefixed(items, "-"),
  ];
}

/**
 * Python `_delivery_story_scope_hypothesis`.
 *
 * The checkout/address special case is product copy, not a heuristic worth
 * generalizing: it is reproduced verbatim because the surface is graded byte for
 * byte.
 */
function deliveryStoryScopeHypothesis(report: NavigatorFlowUnitReport): string {
  const title = pyStrip(report.activeItemTitle ?? "");
  const normalized = title.toLowerCase();
  if (normalized.includes("checkout") && normalized.includes("address")) {
    return (
      "This Delivery Story should let the customer enter checkout, provide or confirm " +
      "a delivery address, and leave the order ready for the next checkout step."
    );
  }
  if (title) return `This Delivery Story should deliver the scope implied by: ${title}.`;
  return (
    "I will infer a complete Delivery Story plan from the roadmap and project context; " +
    "correct or add any scope details before I proceed."
  );
}

/** Python `render_flow_unit_scope_confirmation_report`. */
export function renderFlowUnitScopeConfirmationReport(report: NavigatorFlowUnitReport): string {
  const deliveryStory = report.flowUnit === FLOW_UNIT_DELIVERY_STORY;
  const surfaceName = deliveryStory
    ? "delivery_story_scope_confirmation"
    : "next_story_confirmation";
  const title = deliveryStory
    ? "       🧭  DELIVERY STORY SCOPE CONFIRMATION"
    : "       🧭  NEXT STORY CONFIRMATION";
  const understanding = deliveryStory
    ? deliveryStoryScopeHypothesis(report)
    : "We will continue story by story. The next step is to confirm which child story " +
      "becomes the next Navigator-facing lifecycle unit.";
  const scopeLabel = deliveryStory ? "Work packages in scope" : "Recommended story";
  const scopeItems = deliveryStory
    ? report.cursor.childWorkItems
    : report.cursor.childWorkItems.slice(0, 1);
  const prompt = deliveryStory
    ? "Before I create the DS Plan, correct or add anything:"
    : "Before I create the Story Plan, correct or add anything:";
  const question = deliveryStory
    ? "1. Is this the right scope?"
    : "1. Is this the right next story?";
  const body = [
    "Delivery",
    "",
    "╭────────────────────────────────────────────────────────╮",
    cardText(title),
    "│                                                        │",
    cardText("My understanding"),
    ...cardWrapped(understanding),
    "│                                                        │",
    cardText("Active delivery"),
    ...cardWrapped(activeDelivery(report)),
    ...scopeItemLines(scopeLabel, scopeItems),
    "│                                                        │",
    ...cardWrapped(prompt),
    ...cardWrapped(question),
    "│                                                        │",
    ...cardWrapped("Choose the next move when ready."),
    "╰────────────────────────────────────────────────────────╯",
  ];
  return wrapAriadSurface(surfaceName, `${body.join("\n")}\n`);
}

/** Python `_flow_unit_change_summary` / `_description` / `_next_movement`. */
function flowUnitChangeSummary(flowUnit: string): string {
  return flowUnit === FLOW_UNIT_DELIVERY_STORY
    ? "Navigator-facing lifecycle will continue at Delivery Story level."
    : "Navigator-facing lifecycle will continue story by story.";
}

function flowUnitDescription(flowUnit: string): string {
  return flowUnit === FLOW_UNIT_DELIVERY_STORY
    ? "the Delivery Story becomes the lifecycle unit; child stories remain traceable work packages"
    : "child stories get their own Navigator checkpoints";
}

function flowUnitNextMovement(flowUnit: string): string {
  return flowUnit === FLOW_UNIT_DELIVERY_STORY
    ? "Plan the Delivery Story as the Navigator-facing unit."
    : "Confirm the recommended child story, then Plan.";
}

/** Python `render_navigator_flow_unit_report`: the READ face. */
export function renderNavigatorFlowUnitReport(report: NavigatorFlowUnitReport): string {
  const body = [
    "Delivery",
    "",
    "╭────────────────────────────────────────────────────────╮",
    "│        🧭  FLOW UNIT SELECTED                         │",
    "│                                                        │",
    cardText("What changed?"),
    ...cardWrapped(flowUnitChangeSummary(report.flowUnit)),
    "│                                                        │",
    cardText("Active delivery"),
    ...cardWrapped(activeDelivery(report)),
    "│                                                        │",
    cardText("Selected flow unit"),
    cardText(report.flowUnit),
    ...cardWrapped(flowUnitDescription(report.flowUnit)),
    "│                                                        │",
    cardText("Next movement"),
    ...cardWrapped(flowUnitNextMovement(report.flowUnit)),
    "│                                                        │",
    ...cardWrapped("Choose the next move when ready."),
    "╰────────────────────────────────────────────────────────╯",
  ];
  return wrapAriadSurface("navigator_flow_unit", `${body.join("\n")}\n`);
}
